### DISCUSSION TITLE
Issue with the length(m) in GPR files.

### DISCUSSION THREAD

**Discussion by zamanimd** [emanuelhuber/RGPR#61] [Q&A] [answered]
Hi,

Thanks Emanuel Huber for RGPR.

I have an issue with the length(m) in GPR files.
For example, I have a GPR file with a 127 m length but in RGPR it shows 0.0017 m.

I want to resolve that issue.

Thanks!

**Comment by emanuelhuber** [comment, marked as answer]
Hello @zamanimd 

Thank you for your message!

I suggest you to first try to re-install RGPR and re-read your file.

If the issue persists, it may be due to geographical coordinates. You can check the coordinate reference system (CRS) with the `crs()` function. Le me know what is the CRS of your data.

Else, you can try to read your data by setting the argument `interp_pos = FALSE`:

```
x <- readGPR("your_file_name.dt1", interp_pos = FALSE)
plot(x)
```

Let me know if it solves your problem.

Best,

Emanuel

**Comment by zamanimd** [comment]
Hi,
Thanks for your response.

My file was exported from Prism2 in SGY format, and this is my file information:

*** Class GPR ***
 name        = mtm-7apr-2-01
 filepath    = mtm-7apr-2-01.sgy
 description = 
 survey date = 2023-03-06
 reflection, 0 MHz, Window length = 799.715 ns, dz = 1.565 ns
 329 traces, 0.00174541121789117 m
 ****************

This is CRS() of my data:
[1] "+init=epsg:4326"

I use this code  
`x <- readGPR("your_file_name.dt1", interp_pos = FALSE)`
And I get this result:

*** Class GPR ***
 name        = mtm-7apr-2-01
 filepath    = mtm-7apr-2-01.sgy
 description = 
 survey date = 2023-03-06
 reflection, 0 MHz, Window length = 799.715 ns, dz = 1.565 ns
 329 traces, 328 m
 ****************
 
 Thanks.

**Reply by emanuelhuber** [reply]
Great!
Note that you can project the trace coordinates in another coordinate reference system (CRS) with the function `trProject`, e.g.:

```
x <- reagGPR("my_file_name.dt1", interp_pos = FALSE)
crs(x)
x1 <- trProject(x, "+init=epsg:2056")  # swiss coordinate system
```

Best,

Emanuel

**Reply by zamanimd** [reply]
Hi Emanuel,

I did that and I got 200m after using `trProject`
why do I have to do that?

Then I try to use `interpSlices`
`SXY <- interpSlices(SU, dx = 0.05, dy = 0.05, dz = 0.05, h = 6)" and I have this error`

Error: cannot allocate vector of size 409.4 Gb

Thanks for your help.

**Reply by zamanimd** [reply]
I got results by changing the `dx` and `dy` scale from 0.05 to 0.5

**Reply by zamanimd** [reply]
I want to know how to export GPRcube as a file for use in the other application. (for example Grass or QGIS)

I try to use `writeGPR` but it has error
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘writeGPR’ for signature ‘"GPRcube"’

**Reply by emanuelhuber** [reply]
You can convert the `GPRcube` object to a `raster` object (from the raster package) and then export the raster to .tif file.

Check this section of the tutorial: https://emanuelhuber.github.io/RGPR/05_RGPR_tutorial_GPR-data-time-slice-interpolation-3D/#export-slices-as-raster

Maybe you need to first load the package `raster` with `library(raster)`.

Note that may have to specify the format (`GTiff` for spatial Tiff) and specify `overwrite=TRUE` if you have to re-export the slices.

```
raster::writeRaster(r, filename = "slice10.tif",  format="GTiff", overwrite=TRUE)
```

Does it solve your problem?

Best,

Emanuel

rf <- writeRaster(r, filename=file.path(tmp, "test.tif"), format="GTiff", overwrite=TRUE)

**Reply by emanuelhuber** [reply]
> Hi Emanuel,
> 
> I did that and I got 200m after using `trProject` why do I have to do that?
> 

Here some detailed information about CRS.

Your CRS is the epsg code "4326", that is a geographic CRS where the position is defined by latitude and longitude (check this website: https://epsg.io/4326).

A geographic CRS is great when you work with data that are located all over the world.

When you work with spatial data within a specific region (like a country), it is better to work with a projected CRS where the position is defined in meters (and no more in latitude/longitude).

Check here to learn more about the difference between geographic and projected CRS: https://www.esri.com/arcgis-blog/products/arcgis-pro/mapping/coordinate-systems-difference/

I do not know where your data come from, but check there, what is the commonly used projected CRS. You can use this site to check that too and to get the epsg code: https://epsg.io
Then, you can project your data into this CRS with the function `trProject()` before you interpolate slices.

Does it make sense?

Best,

Emanuel

**Reply by zamanimd** [reply]
Thanks for your messages.

Based on your tips about the coordination system, I fixed the length of my files And Export each slice as a raster.
I want to know, Is it possible for exporting GPRcube as an object (all slices together)?

For example, I tried to use this code but it has an error:
`rx <- as.raster(SXY[,,])`
`rf <- writeRaster(rx, filename="GPRcube.tif", format="GTiff", overwrite=TRUE)`

Error in .startGDALwriting(x, filename, gdal = options, ...) : 
  no slot of name "file" for this object of class "RasterStack"
  
Actually, I would like to export GPRcube as a 3D object for processing data in the next step of my processing.

**Reply by emanuelhuber** [reply]
Good question!
I did some testing and found out, that tif files allow only 3 layers to be saved. 

To save all the layers in one GIS-file, use the NetCDF format. You need first to install the `ncdf4` package:

```r
install.packages("ncdf4")
r <- as.raster(SXY)
raster::writeRaster(r, filename = "all_slices.nc", format="CDF")
```

Then, you can read the file in QGIS, for example. There, you have to activate the "group" you are interested in (see below).
Let me know if it helps you!

Best,

Emanuel

![image](https://user-images.githubusercontent.com/12541363/226467119-41117306-658a-42b7-8221-28c21ad816fc.png)

**Reply by zamanimd** [reply]
Hi Emanuel,
Thanks for your time.
That was so useful.

I used this one for exporting data as a raster file in tiff format too.

```
r_stack <- as.raster(SXY)
r_brick <- brick(r_stack)
writeRaster(r_brick, "all_slices.tif", format = "GTiff", overwrite = TRUE)
```

Best,
Mohammad
